File: code/src/Probabilities.cpp
#include "Probabilities.h"
#include <eigen3/Eigen/Eigenvalues>
#include <complex>

const std::complex<double> ProbConst::I = std::complex<double>(0, 1);
const double ProbConst::hbar = 6.58211928*1.e-25;
const double ProbConst::clight = 299792458;
//double GevkmToevsq = hbar*clight*1.e15;
const double ProbConst::GevkmToevsq = 0.197327; // Valor aproximado

Probabilities::Probabilities() {

}

Probabilities::~Probabilities() {

}

void Probabilities::StandardOscilation(
	double Energ, int sigN, double L, double rhomat, double* th, 
    double dCP, double* dm, double* alpha, double** P) {
	InvisibleDecay(Energ, sigN, L, rhomat, th, dCP, dm, alpha, P);	
}

void Probabilities::InvisibleDecay(
    double Energ, int sigN, double L, double rhomat, double* th, 
    double dCP, double* dm, double* alpha, double** P) {
	Eigen::MatrixXcd UPMNS(3, 3);
	Eigen::MatrixXcd Hd(3, 3);
	Eigen::MatrixXcd Pot(3, 3);
	Eigen::MatrixXcd Hff(3, 3);
	Eigen::MatrixXcd S(3, 3);
	Eigen::MatrixXcd V(3, 3);

	double En = Energ * 1e9;
	double th12 = th[0];
	double th13 = th[1];
	double th23 = th[2];
	double delta = sigN * dCP;
	double dm21 = dm[0];
	double dm31 = dm[1];
	double alpha1 = alpha[0];
	double alpha2 = alpha[1];
	double alpha3 = alpha[2];
	double rho = sigN * rhomat;
	// Definir Matrices para cálculos
    std::complex<double> U[3][3];
    std::complex<double> DM[3][3];

	/* Compute vacuum mixing matrix */
	U[0][0] = cos(th12) * cos(th13);
	U[0][1] = sin(th12) * cos(th13);
	U[0][2] = sin(th13) * exp(-ProbConst::I * delta);

	U[1][0] = -sin(th12) * cos(th23) - cos(th12) * sin(th23) * sin(th13) * exp(ProbConst::I * delta);
	U[1][1] = cos(th12) * cos(th23) - sin(th12) * sin(th23) * sin(th13) * exp(ProbConst::I * delta);
	U[1][2] = sin(th23) * cos(th13);

	U[2][0] = sin(th12) * sin(th23) - cos(th12) * cos(th23) * sin(th13) * exp(ProbConst::I * delta);
	U[2][1] = -cos(th12) * sin(th23) - sin(th12) * cos(th23) * sin(th13) * exp(ProbConst::I * delta);
	U[2][2] = cos(th23) * cos(th13);

	/* Matriz de masas y Decay */
	DM[0][0] = std::complex<double>(0, -0.5 * alpha1 / En);
	DM[0][1] = std::complex<double>(0, 0);
	DM[0][2] = std::complex<double>(0, 0);

	DM[1][0] = std::complex<double>(0, 0);
	DM[1][1] = std::complex<double>(0.5 * dm21 / En, -0.5 * alpha2 / En);
	DM[1][2] = std::complex<double>(0, 0);

	DM[2][0] = std::complex<double>(0, 0);
	DM[2][1] = std::complex<double>(0, 0);
	DM[2][2] = std::complex<double>(0.5 * dm31 / En, -0.5 * alpha3 / En);

	/* Inicializando las matrices para Eigen */
	UPMNS << U[0][0], U[0][1], U[0][2],
		U[1][0], U[1][1], U[1][2],
		U[2][0], U[2][1], U[2][2];
	Hd << DM[0][0], DM[0][1], DM[0][2],
		DM[1][0], DM[1][1], DM[1][2],
		DM[2][0], DM[2][1], DM[2][2];
	Pot << std::complex<double>(rho * 7.63247 * 0.5 * 1.e-14, 0), DM[0][1], DM[0][2],
		DM[1][0], DM[0][0], DM[1][2],
		DM[2][0], DM[2][1], DM[0][0];

	/* Hamiltoniano final efectivo */
	Hff = UPMNS * Hd * UPMNS.adjoint() + Pot;

	/* Calculando los autovalores y autovectores */
	Eigen::ComplexEigenSolver<Eigen::MatrixXcd> tmp;
	tmp.compute(Hff);

	/* Calculamos la matriz S y ordenamos los autovalores */
	V = tmp.eigenvectors();
	S << exp(-ProbConst::I * tmp.eigenvalues()[0] * L * 1.e9 / ProbConst::GevkmToevsq), DM[0][0], DM[0][0],
		DM[0][0], exp(-ProbConst::I * tmp.eigenvalues()[1] * L * 1.e9 / ProbConst::GevkmToevsq), DM[0][0],
		DM[0][0], DM[0][0], exp(-ProbConst::I * tmp.eigenvalues()[2] * L * 1.e9 / ProbConst::GevkmToevsq);

	S = (V)*S * (V.inverse());

	// Calculando la matriz de probabilidad
	for (int i = 0; i < 3; i++)
	{
		for (int j = 0; j < 3; j++)
		{
			P[i][j] = abs(S.col(i)[j] * S.col(i)[j]);
		}
	} // Fila j , columna i
}

void Probabilities::NonStandardInteraction(
    double Energ, int sigN, double L, double rhomat, double* th, 
    double dCP, double* dm, double* parmNSI, double** P) {
    Eigen::MatrixXcd UPMNS(3, 3);
    Eigen::MatrixXcd Hd(3, 3);
    Eigen::MatrixXcd Pot(3, 3);
    Eigen::MatrixXcd MNSI(3, 3);
    Eigen::MatrixXcd Hff(3, 3);
    Eigen::MatrixXcd S(3, 3);
    Eigen::MatrixXcd V(3, 3);

	double En = Energ*1e9;
	double th12 = th[0];
	double th13 = th[1];
	double th23 = th[2];
	double delta = sigN*dCP;
	double dm21 = dm[0];
	double dm31 = dm[1];
	double rho = sigN*rhomat;
    // Definir Matrices para cálculos
    std::complex<double> U[3][3];
    std::complex<double> DM[3][3];
    std::complex<double> NSI[3][3];

    /* Compute vacuum mixing matrix */
    U[0][0] = cos(th12)*cos(th13);
    U[0][1] = sin(th12)*cos(th13);
    U[0][2] = sin(th13) * exp(-ProbConst::I*delta);

    U[1][0] = -sin(th12)*cos(th23) - cos(th12)*sin(th23)*sin(th13) * exp(ProbConst::I*delta);
    U[1][1] =  cos(th12)*cos(th23) - sin(th12)*sin(th23)*sin(th13) * exp(ProbConst::I*delta);
    U[1][2] =  sin(th23)*cos(th13);

    U[2][0] =  sin(th12)*sin(th23) - cos(th12)*cos(th23)*sin(th13) * exp(ProbConst::I*delta);
    U[2][1] = -cos(th12)*sin(th23) - sin(th12)*cos(th23)*sin(th13) * exp(ProbConst::I*delta);
    U[2][2] =  cos(th23)*cos(th13);

    /* Matriz de masas */
    DM[0][0] = std::complex<double>(0, 0);
    DM[0][1] = std::complex<double>(0, 0);
    DM[0][2] = std::complex<double>(0, 0);

    DM[1][0] = std::complex<double>(0, 0);
    DM[1][1] = std::complex<double>(0.5*dm21/En, 0);
    DM[1][2] = std::complex<double>(0, 0);

    DM[2][0] = std::complex<double>(0, 0);
    DM[2][1] = std::complex<double>(0, 0);
    DM[2][2] = std::complex<double>(0.5*dm31/En, 0);
    
    /* Matriz de parámetros NSI */
    NSI[0][0] = parmNSI[0];
    NSI[0][1] = parmNSI[3] * exp(ProbConst::I*parmNSI[4]);
    NSI[0][2] = parmNSI[5] * exp(ProbConst::I*parmNSI[6]);

    NSI[1][0] = parmNSI[3] * exp(-ProbConst::I*parmNSI[4]);
    NSI[1][1] = parmNSI[1];
    NSI[1][2] = parmNSI[7] * exp(ProbConst::I*parmNSI[8]);

    NSI[2][0] = parmNSI[5] * exp(-ProbConst::I*parmNSI[6]);
    NSI[2][1] = parmNSI[7] * exp(-ProbConst::I*parmNSI[8]);
    NSI[2][2] = parmNSI[2];

	/* Inicializando las matrices para Eigen */
	UPMNS <<
		U[0][0], U[0][1], U[0][2],
		U[1][0], U[1][1], U[1][2], 
		U[2][0], U[2][1], U[2][2];
	Hd <<
		DM[0][0], DM[0][1], DM[0][2],
		DM[1][0], DM[1][1], DM[1][2], 
		DM[2][0], DM[2][1], DM[2][2];
	Pot <<
		std::complex<double>(rho*7.63247*0.5*1.e-14, 0), DM[0][1], DM[0][2],
		DM[1][0], DM[0][0], DM[1][2], 
		DM[2][0], DM[2][1], DM[0][0];
	MNSI <<
		NSI[0][0], NSI[0][1], NSI[0][2],
		NSI[1][0], NSI[1][1], NSI[1][2], 
		NSI[2][0], NSI[2][1], NSI[2][2];

	/* Hamiltoniano final efectivo */
	Hff = UPMNS * Hd * UPMNS.adjoint() + rho*7.63247*0.5*1.e-14*MNSI + Pot ;

	/* Calculando los autovalores y autovectores */
	Eigen::ComplexEigenSolver<Eigen::MatrixXcd> tmp;
	tmp.compute(Hff);

	/* Calculamos la matriz S y ordenamos los autovalores */
	V = tmp.eigenvectors() ;
	S <<
		exp(-ProbConst::I*tmp.eigenvalues()[0]*L*1.e9/ProbConst::GevkmToevsq), DM[0][0], DM[0][0],
		DM[0][0], exp(-ProbConst::I*tmp.eigenvalues()[1]*L*1.e9/ProbConst::GevkmToevsq), DM[0][0], 
		DM[0][0], DM[0][0], exp(-ProbConst::I*tmp.eigenvalues()[2]*L*1.e9/ProbConst::GevkmToevsq);

	S = ( V ) * S * ( V.inverse() );

	// Calculando la matriz de probabilidad
	for(int i=0; i<3; i++){
		for(int j=0; j<3; j++) {
			P[i][j] = abs(S.col(i)[j]*S.col(i)[j]); 
		} 
	} // Fila j , columna i
}

void Probabilities::ViolationPrincipleDecay(
	double Energ, int sigN, double L, double rhomat, double* th, 
	double dCP, double* dm, double* gamma, double** P) {
	Eigen::MatrixXcd UPMNS(3, 3);
	Eigen::MatrixXcd Hd(3, 3);
	Eigen::MatrixXcd Pot(3, 3);
	Eigen::MatrixXcd Hff(3, 3);
	Eigen::MatrixXcd S(3, 3);
	Eigen::MatrixXcd V(3, 3);

	double En = Energ*1e9;
	double th12 = th[0];
	double th13 = th[1];
	double th23 = th[2];
	double delta = sigN*dCP;
	double dm21 = dm[0];
	double dm31 = dm[1];
	double gamma1 = gamma[0];
	double gamma2 = gamma[1];
	double gamma3 = gamma[2];
	double rho = sigN*rhomat;
	std::complex<double> U[3][3];
    std::complex<double> DM[3][3];
	/* Compute vacuum mixing matrix */
	U[0][0] = cos(th12)*cos(th13);
	U[0][1] = sin(th12)*cos(th13);
	U[0][2] = sin(th13) * exp(-ProbConst::I*delta);

	U[1][0] = -sin(th12)*cos(th23) - cos(th12)*sin(th23)*sin(th13) * exp(ProbConst::I*delta);
	U[1][1] =  cos(th12)*cos(th23) - sin(th12)*sin(th23)*sin(th13) * exp(ProbConst::I*delta);
	U[1][2] =  sin(th23)*cos(th13);

	U[2][0] =  sin(th12)*sin(th23) - cos(th12)*cos(th23)*sin(th13) * exp(ProbConst::I*delta);
	U[2][1] = -cos(th12)*sin(th23) - sin(th12)*cos(th23)*sin(th13) * exp(ProbConst::I*delta);
	U[2][2] =  cos(th23)*cos(th13);

	/* Matriz de masas y VEP con U_g = U */
	DM[0][0] = std::complex<double>(0+2*En*gamma1, 0);
	DM[0][1] = std::complex<double>(0, 0);
	DM[0][2] = std::complex<double>(0, 0);

	DM[1][0] = std::complex<double>(0, 0);
	DM[1][1] = std::complex<double>(0.5*dm21/En + 2*En*gamma2, 0);
	DM[1][2] = std::complex<double>(0, 0);

	DM[2][0] = std::complex<double>(0, 0);
	DM[2][1] = std::complex<double>(0, 0);
	DM[2][2] = std::complex<double>(0.5*dm31/En + 2*En*gamma3, 0);

	/* Inicializando las matrices para Eigen */
	UPMNS <<
		U[0][0], U[0][1], U[0][2],
		U[1][0], U[1][1], U[1][2], 
		U[2][0], U[2][1], U[2][2];
	Hd <<
		DM[0][0], DM[0][1], DM[0][2],
		DM[1][0], DM[1][1], DM[1][2], 
		DM[2][0], DM[2][1], DM[2][2];
	Pot <<
		std::complex<double>(rho*7.63247*0.5*1.e-14, 0), DM[0][1], DM[0][2],
		DM[1][0], DM[0][0], DM[1][2], 
		DM[2][0], DM[2][1], DM[0][0];

	/* Hamiltoniano final efectivo */
	Hff = UPMNS * Hd * UPMNS.adjoint() + Pot ;

	/* Calculando los autovalores y autovectores */
	Eigen::ComplexEigenSolver<Eigen::MatrixXcd> tmp;
	tmp.compute(Hff);

	/* Calculamos la matriz S y ordenamos los autovalores */
	V = tmp.eigenvectors() ;
	S <<
		exp(-ProbConst::I*tmp.eigenvalues()[0]*L*1.e9/ProbConst::GevkmToevsq), DM[0][0], DM[0][0],
		DM[0][0], exp(-ProbConst::I*tmp.eigenvalues()[1]*L*1.e9/ProbConst::GevkmToevsq), DM[0][0], 
		DM[0][0], DM[0][0], exp(-ProbConst::I*tmp.eigenvalues()[2]*L*1.e9/ProbConst::GevkmToevsq);

	S = ( V ) * S * ( V.inverse() );

	// Calculando la matriz de probabilidad
	for(int i=0; i<3; i++) {
		for(int j=0; j<3; j++) {
			P[i][j] = abs(S.col(i)[j]*S.col(i)[j]);
		}
	}
}

File: code/python/src/probabilities.py
from time import process_time_ns
import numpy as np
from scipy.sparse.linalg import LinearOperator
from integrate.gaussLegendre import NintegrGLQ
from utils import ProbConst, PhiIntegration, matrixAdjoint

def InvisibleDecay(_U, _energy, _sigN, _L, _rho, _dm, _alpha, _P):

    Pot = np.ndarray((3,3),dtype=np.complex)
    Hff = np.ndarray((3,3),dtype=np.complex)
    S = np.ndarray((3,3),dtype=np.complex)
    V = np.ndarray((3,3),dtype=np.complex)

    energy = _energy * 1e9
    rho = _sigN * _rho

    DM = np.zeros((3,3),dtype=np.complex)

    DM[0,0] = np.complex(0, -0.5 * _alpha[0] / energy)
    DM[1,1] = np.complex(0.5 * _dm[0] / energy, -0.5 * _alpha[1] / energy)
    DM[2,2] = np.complex(0.5 * _dm[1] / energy, -0.5 * _alpha[2] / energy)

    Pot[0,0] = np.complex(rho * 7.63247 * 0.5 * 1e-14, 0)
    Pot[0,1] = DM[0][1]
    Pot[0,2] = DM[0][2]
    Pot[1,0] = DM[1][0]
    Pot[1,1] = DM[0][0]
    Pot[1,2] = DM[1][2]
    Pot[2,0] = DM[2][0]
    Pot[2,1] = DM[2][1]
    Pot[2,2] = DM[0][0]

    UPMNS = np.copy(_U)
    HD = np.copy(DM)
    T = np.copy(matrixAdjoint(UPMNS))

    N = UPMNS @ HD
    A = N @ T

    Hff = A + Pot

    eigvalues, eigvectors = np.linalg.eig(Hff)

    V = eigvectors

    S[0,0] = np.exp(-ProbConst.I * eigvalues[0] * _L * 1.e9 / ProbConst.GevkmToevsq)
    S[0,1] = DM[0][0]
    S[0,2] = DM[0][0]
    S[1,0] = DM[0][0]
    S[1,1] = np.exp(-ProbConst.I * eigvalues[1] * _L * 1e9 / ProbConst.GevkmToevsq)
    S[1,2] = DM[0][0]
    S[2,0] = DM[0][0]
    S[2,1] = DM[0][0]
    S[2,2] = np.exp(-ProbConst.I * eigvalues[2] * _L * 1e9 / ProbConst.GevkmToevsq)

    S = (V) @ S @ (np.linalg.inv(V))

    for i in range(3):
        for j in range(3):
            _P[i][j] = abs(S[:,i][j] * S[:,i][j])
    

def StandardOscilation(_U, _energy, _sigN, _L, _rho, _dm, _alpha, _P):
    
    InvisibleDecay(_U, _energy, _sigN, _L, _rho, _dm, _alpha, _P)

def NonStandardInteraction(_U, _energy, _sigN, _L, _rho, _dm, _parmNSI, _P):
    # hermitian matrix

    Pot = np.ndarray((3,3), dtype=np.complex)
    MNSI = np.ndarray((3,3), dtype=np.complex)
    Hff = np.ndarray((3,3), dtype=np.complex)
    S = np.ndarray((3,3), dtype=np.complex)
    V = np.ndarray((3,3), dtype=np.complex)

    energy = _energy * 1e9
    rho = _sigN * _rho

    DM = np.zeros((3,3), dtype=np.complex)
    NSI = np.ndarray((3,3), dtype=np.complex)

    DM[1,1] = np.complex(0.5 * _dm[0] / energy, 0)
    DM[2,2] = np.complex(0.5 * _dm[1] / energy, 0)

    NSI[0,0] = _parmNSI[0]
    NSI[0,1] = _parmNSI[3] * np.exp(ProbConst.I * _parmNSI[4])
    NSI[0,2] = _parmNSI[5] * np.exp(ProbConst.I * _parmNSI[6])
    NSI[1,0] = _parmNSI[3] * np.exp(-ProbConst.I * _parmNSI[4])
    NSI[1,1] = _parmNSI[1]
    NSI[1,2] = _parmNSI[7] * np.exp(ProbConst.I * _parmNSI[8])
    NSI[2,0] = _parmNSI[5] * np.exp(-ProbConst.I * _parmNSI[6])
    NSI[2,1] = _parmNSI[7] * np.exp(-ProbConst.I * _parmNSI[8])
    NSI[2,2] = _parmNSI[2]

    Pot[0,0] = np.complex(rho * 7.63247 * 0.5 * 1.e-14, 0)
    Pot[0,1] = DM[0][1]
    Pot[0,2] = DM[0][2]
    Pot[1,0] = DM[1][0]
    Pot[1,1] = DM[0][0]
    Pot[1,2] = DM[1][2]
    Pot[2,0] = DM[2][0]
    Pot[2,1] = DM[2][1]
    Pot[2,2] = DM[0][0]

    MNSI = np.copy(NSI)

    UPMNS = np.copy(_U)

    HD = np.copy(DM)

    N = UPMNS @ HD
    T = np.copy(matrixAdjoint(UPMNS))
    A = N @ T

    Hff = A + (rho * 7.63247 * 0.5 * 1.e-14 * MNSI) + Pot

    eigvalues, eigvectors = np.linalg.eigh(Hff)

    V = eigvectors

    S[0,0] = np.exp(-ProbConst.I * eigvalues[0] * _L * 1.e9 / ProbConst.GevkmToevsq)
    S[0,1] = DM[0][0]
    S[0,2] = DM[0][0]
    S[1,0] = DM[0][0]
    S[1,1] = np.exp(-ProbConst.I * eigvalues[1] * _L * 1.e9 / ProbConst.GevkmToevsq)
    S[1,2] = DM[0][0]
    S[2,0] = DM[0][0]
    S[2,1] = DM[0][0]
    S[2,2] = np.exp(-ProbConst.I * eigvalues[2] * _L * 1.e9 / ProbConst.GevkmToevsq)
    S = (V) @ S @ (np.linalg.inv(V))

    for i in range(3):
        for j in range(3):
            _P[i][j] = abs(S[:,i][j] * S[:,i][j])

def ViolationPrincipleDecay(_U, _energy, _sigN, _L, _rho, _dm, _gamma, _P):
    # hermitian matrix

    Pot = np.ndarray((3,3), dtype=np.complex)
    Hff = np.ndarray((3,3), dtype=np.complex)
    S = np.ndarray((3,3), dtype=np.complex)
    V = np.ndarray((3,3), dtype=np.complex)

    energy = _energy * 1.e9
    rho = _sigN * _rho

    DM = np.zeros((3,3), dtype=np.complex)

    DM[0,0] = np.complex(2 * energy * _gamma[0], 0)
    DM[1,1] = np.complex(0.5 * _dm[0] / energy + 2 * energy * _gamma[1], 0)
    DM[2,2] = np.complex(0.5 * _dm[1] / energy + 2 * energy * _gamma[2], 0)

    Pot[0,0] = np.complex(rho * 7.63247 * 0.5 * 1e-14, 0)
    Pot[0,1] = DM[0][1]
    Pot[0,2] = DM[0][2]
    Pot[1,0] = DM[1][0]
    Pot[1,1] = DM[0][0]
    Pot[1,2] = DM[1][2]
    Pot[2,0] = DM[2][0]
    Pot[2,1] = DM[2][1]
    Pot[2,2] = DM[0][0]

    UPMNS = np.copy(_U)

    HD = np.copy(DM)

    N = UPMNS @ HD
    T = np.copy(matrixAdjoint(UPMNS))
    A = N @ T

    Hff = A + Pot

    eigvalues, eigvectors = np.linalg.eigh(Hff)

    V = eigvectors

    S[0,0] = np.exp(-ProbConst.I * eigvalues[0] * _L * 1e9 / ProbConst.GevkmToevsq)
    S[0,1] = DM[0][0]
    S[0,2] = DM[0][0]
    S[1,0] = DM[0][0]
    S[1,1] = np.exp(-ProbConst.I * eigvalues[1] * _L * 1e9 / ProbConst.GevkmToevsq)
    S[1,2] = DM[0][0]
    S[2,0] = DM[0][0]
    S[2,1] = DM[0][0]
    S[2,2] = np.exp(-ProbConst.I * eigvalues[2] * _L * 1e9 / ProbConst.GevkmToevsq)

    S = (V) @ S @ (np.linalg.inv(V))

    for i in range(3):
        for j in range(3):
            _P[i][j] = abs(S[:,i][j] * S[:,i][j])

def ProbabilityVis(_energy, _L, _rho, _th, _dm, d, _alpha, _mlight, _tfi, _tsi, _tff, _tsf, _tpar, _thij, _tqcoup):

    Pot = np.ndarray((3,3), dtype=np.complex)

    Enf = _energy * 1.e9
    mss = np.ndarray((3),  dtype=np.complex)
    
    if (_dm[1] > 0):
        mss[0] = _mlight
        mss[1] = _dm[0] + _mlight
        mss[2] = _dm[1] + _mlight
    else:
        mss[0] = _mlight
        mss[1] = -_dm[1] + _mlight
        mss[2] = _dm[0] - _dm[1] + _mlight
    
    mm = min(20, (mss[_tpar] / mss[_thij]) * _energy)

    Pot[0,0] = np.complex(_rho * 7.63247 * 0.5 * 1e-14, 0)
    Pot[0,1] = ProbConst.Z0
    Pot[0,2] = ProbConst.Z0
    Pot[1,0] = ProbConst.Z0
    Pot[1,1] = ProbConst.Z0
    Pot[1,2] = ProbConst.Z0
    Pot[2,0] = ProbConst.Z0
    Pot[2,1] = ProbConst.Z0
    Pot[2,2] = ProbConst.Z0
    return np.real(1.e9 * NintegrGLQ(lambda x: PhiIntegration(x, mss, Pot, Enf, _L, _th, d, _alpha, _tfi, _tsi, _tff, _tsf, _tpar, _thij, _tqcoup), (1.00001) * _energy, mm))